source/glkernel/include/glkernel/noise.hpp

```diff
@@ -131,6 +131,192 @@ void normal(tkernel<V<T, P>> & kernel, const V<T, P> & mean, const V<T, P> & std
+    const auto aba = glm::dot(grad3<T>(is + 0, it + 1, iu + 0, r), f - glm::tvec3<T, glm::highp>(0., 1., 0.));
+    const auto bba = glm::dot(grad3<T>(is + 1, it + 1, iu + 0, r), f - glm::tvec3<T, glm::highp>(1., 1., 0.));
+    const auto aab = glm::dot(grad3<T>(is + 0, it + 0, iu + 1, r), f - glm::tvec3<T, glm::highp>(0., 0., 1.));
+    const auto bab = glm::dot(grad3<T>(is + 1, it + 0, iu + 1, r), f - glm::tvec3<T, glm::highp>(1., 0., 1.));
+    const auto abb = glm::dot(grad3<T>(is + 0, it + 1, iu + 1, r), f - glm::tvec3<T, glm::highp>(0., 1., 1.));
+    const auto bbb = glm::dot(grad3<T>(is + 1, it + 1, iu + 1, r), f - glm::tvec3<T, glm::highp>(1., 1., 1.));
+
+    const auto i = glm::mix(
+        glm::tvec4<T, glm::highp>(aaa, aab, aba, abb),
+        glm::tvec4<T, glm::highp>(baa, bab, bba, bbb), 
+        fade(f[0]));
+    const auto j = glm::mix(
+        glm::tvec2<T, glm::highp>(i[0], i[1]),
+        glm::tvec2<T, glm::highp>(i[2], i[3]), fade(f[1]));
+
+    return glm::mix(j[0], j[1], fade(f[2]));
+}
+
+} // anonymous namespace
+
+template<typename T, typename std::enable_if<std::is_floating_point<T>::value>::type *>
+void perlin(tkernel<T> & kernel
+    , const T scale
+    , const PerlinNoiseType type
+    , const int startFrequency
+    , const int octaves
+    , const bool normalize)
+{
+    if (kernel.size() < 1)
+        return;
+
+    std::vector<T> fo(octaves);
+
+    for (int o = 0; o < octaves; ++o)
+    {
+        fo[o] = static_cast<T>(1.0 / (1 << o));
+    }
+
+    T minp = scale;
+    T maxp = 0.0;
+
+    int k = 0;
+    for (int z = 0; z < kernel.depth(); ++z)
+    {
+        auto zf = static_cast<T>(z) / kernel.depth();
+        for (int y = 0; y < kernel.height(); ++y)
+        {
+            auto yf = static_cast<T>(y) / kernel.height();
+            for (int x = 0; x < kernel.width(); ++x)
+            {
+                auto xf = static_cast<T>(x) / kernel.width();
+
+                // collect noise values over multiple octaves
+                T p = 0.5f;
+                for (int o = 0; o < octaves; ++o)
+                {
+                    T po = noise3(xf, yf, zf, o + startFrequency);
+                    T pf = fo[o] * po;
+
+                    switch (type)
+                    {
+                    case PerlinNoiseType::Standard:
+                        p += o > 0 ? 0.f : po;
+                        break;
+                    case PerlinNoiseType::Cloud:
+                        p += pf;
+                        break;
+                    case PerlinNoiseType::CloudAbs:
+                        p += fabs(pf);
+                        break;
+                    case PerlinNoiseType::Wood:
+                        p += (pf * 8.f) - static_cast<int>(pf * 8.f);
+                        break;
+                    case PerlinNoiseType::Paper:
+                        p += po * po * (pf > 0 ? 1.f : -1.f);
+                        break;
+                    };
+                }
+
```

Comment:
This is a nesting level of 6? we don't have nesting level above 3! :godmode: - perhaps we add a ```location``` function to ```Kernel``` that allows to retrieve the s, t, and r indices based on a given index ... -> with that the first three loops can be reduced to a single one. Furthermore, the switch case logic should be extracted to an additional function ...